lib/client.js

```diff
@@ -246,20 +282,47 @@ class ServiceClient {
             accept: 'application/json'
         }, params.headers);
 
-        return new Promise((resolve, reject) => this.breaker.run(
-            (success, failure) => {
+        const {
+            retries,
+            factor,
+            minTimeout,
+            maxTimeout,
+            randomize,
+            transientErrorCheck,
+            onRetry
+        } = this.options.retryOptions;
+
+        const opts = {
+            retries,
+            factor,
+            minTimeout,
+            maxTimeout,
+            randomize
+        };
+
+        const operation = retry.operation(opts);
+
+        return new Promise((resolve, reject) => operation.attempt((currentAttempt) => {
+            this.breaker.run((success, failure) => {
                 return requestWithFilters(params, this.options.filters)
                     .then(result => {
-                        success(); resolve(result);
+                        success();
+                        resolve(result);
                     })
                     .catch(error => {
-                        failure(); reject(error);
+                        failure();
+                        const shouldRetry = operation.retry(error);
+                        if (!transientErrorCheck(error) || !shouldRetry) {
+                            reject(error);
+                            return;
+                        }
+                        onRetry(currentAttempt, error);
```

Comment:
Shouldn't it be `currentAttempt + 1`? The retry will be the next attempt.

Reply:
The first time onRetry is called, currentAttempt is 1 which I think is correct